single_kernel_mongo/core/operator.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Abstract Operator for Mongo Related Charms."""
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, ClassVar
+
+from ops.charm import RelationDepartedEvent
+from ops.framework import Object
+from ops.model import Unit
+
+from single_kernel_mongo.config.literals import RoleEnum, Substrates
+from single_kernel_mongo.managers.config import CommonConfigManager
+from single_kernel_mongo.managers.mongo import MongoManager
+from single_kernel_mongo.state.charm_state import CharmState
+
+if TYPE_CHECKING:
```

Comment:
The `TYPE_CHECKING` constant is True when we are running some type checking, but False at runtime.
This allows to do some import that exist only when type checking (with mypy, pyright or whatever type checked you're using). This allows to access the full interface in the class during development and type checking but there's no need to import it at runtime.
In this case, it's useful to avoid circular dependencies at runtime which is sadly very hard to get rid of with the ops API where you need to pass the parent object to every other child object.